src/libs/Violations/useViolations.ts

```diff
@@ -0,0 +1,67 @@
+import {useCallback, useMemo} from 'react';
+import useLocalize from '@hooks/useLocalize';
+import {TransactionViolation, ViolationName} from '@src/types/onyx';
+
+/**
+ * Map from Violation Names to the field where that violation can occur
+ */
+const violationFields: Record<ViolationName, ViolationField> = {
+    perDayLimit: 'amount',
+    maxAge: 'date',
+    overLimit: 'amount',
+    overLimitAttendee: 'amount',
+    overCategoryLimit: 'amount',
+    receiptRequired: 'receipt',
+    missingCategory: 'category',
+    categoryOutOfPolicy: 'category',
+    missingTag: 'tag',
+    tagOutOfPolicy: 'tag',
+    missingComment: 'comment',
+    taxRequired: 'tax',
+    taxOutOfPolicy: 'tax',
+    billableExpense: 'billable',
+};
+
+/**
+ * Names of Fields where violations can occur
+ */
+type ViolationField = 'merchant' | 'amount' | 'category' | 'date' | 'tag' | 'comment' | 'billable' | 'receipt' | 'tax';
+
+type ViolationsMap = Map<ViolationField, TransactionViolation[]>;
+
+export default function useViolations(violations: TransactionViolation[]) {
+    const {translate} = useLocalize();
+
+    const violationsByField = useMemo((): ViolationsMap => {
+        const violationGroups = new Map<ViolationField, TransactionViolation[]>();
+
+        for (const violation of violations) {
+            const field = violationFields[violation.name];
+            const existingViolations = violationGroups.get(field) ?? [];
+            violationGroups.set(field, [...existingViolations, violation]);
+        }
+
+        return violationGroups;
+    }, [violations]);
+
+    const hasViolations = useCallback(
+        (field: ViolationField) => {
```

Comment:
Is there a reason to not take a list of every `field` in the `useViolations` argument and just calculate and memoize it all at once?

Reply:
Yeah I went back and forth on this -- the list is very small, so I don't know if the memoization is a net benefit. BUt I'll give it a try.